Some changes to support OpenBSD:

buzzdeee opened the pull request:
* a couple of OpenBSD specific additions to params.pp, i.e. user,
   download tool, paths, service provider, etc.
 * add a $service_flags parameter to instance.pp, since the OpenBSD
   service type supports flags, and those should be settable.
 * add a $service_flags parameter to service.pp, that gets passed
   in from instance.pp, and handed over to openbsd.pp service.
 * add an easy openbsd service provider, that only knows about
   status, ensure, pid_dir, service_flags and init_template
   parameters. There is no "standard" defaults file on OpenBSD.
 * because the defaults_location on OpenBSD is not set, guard
   the augeas call in config.pp to manage the defaults file.
 * added a template for the openbsd service, based on the
   rc script, that comes with the package.


tested with elasticsearch-1.7.2, configured and fired up an instance as well as plugin installation works.

Let me know if there is something that needs change/enhancement to get merged.

thanks,
Sebastian

elasticsearch-release commented:
Jenkins standing by to test this. If you aren't a maintainer, you can ignore this comment. Someone with commit access, please review this and clear it for Jenkins to run; then  say 'jenkins, test it'.

elasticsearch-release commented:
Jenkins standing by to test this. If you aren't a maintainer, you can ignore this comment. Someone with commit access, please review this and clear it for Jenkins to run; then  say 'jenkins, test it'.

electrical commented:
Hi,

Thank you very much for adding the support! I'll have a good look at this very soon :-)

electrical commented on manifests/instance.pp:
```diff
@@ -306,6 +310,7 @@
   elasticsearch::service { $name:
     ensure             => $ensure,
     status             => $status,
+    service_flags      => $init_flags,
```
You have `service_flags` earlier but here its called `init_flags` ?

buzzdeee replied:
indeed, I initially had this as $init_flags, because of all the other init_* variables, but then, since the others don't really make sense that much, I changed it to service_flags, and seem to have this one forgotten, fixed now.

electrical commented on templates/etc/init.d/elasticsearch.OpenBSD.erb:
```diff
@@ -0,0 +1,26 @@
+# $OpenBSD: elasticsearch.rc,v 1.2 2015/07/15 23:34:49 ajacoutot Exp $
+
+daemon="/usr/local/elasticsearch/bin/elasticsearch"
+daemon_flags="-d -Des.config=/etc/elasticsearch/<%= @name %>/elasticsearch.yml -p <%= @pid_dir %>/elasticsearch-<%= @name %>.pid"
```
I think i'm missing a few options here.
from the redhat init script:

```
-Des.default.path.home=$ES_HOME 
-Des.default.path.logs=$LOG_DIR 
-Des.default.path.data=$DATA_DIR 
-Des.default.path.work=$WORK_DIR 
-Des.default.path.conf=$CONF_DIR
```

Could you perhaps include the same options?

buzzdeee replied:
I only have the few parameters in the file for the daemon_flags, since this is what the "original" rc script from the package sets. Therefore that is all 
Further, there is no concept of /etc/{sysconfig|defaults}/progname or the like like on many (or most) Linux distributions. 
That's why I protected the augeas call in config.pp "${elasticsearch::params::defaults_location}/elasticsearch" to only get evaluated, when that defaults_location parameter is set, which it isn't for OpenBSD.
Also I ripped all the defaults_file related stuff out of service/openbsd.pp.
However, the $service_flags parameter of each instance, the user of the module can override the daemon_flags of each rc script.

i.e. 
service { 'elasticsearch_es01':
  flags => '-d -Des.config=/etc/elasticsearch/es01/elasticsearch.yml -p /var/run/elasticsearch/elasticsearch-es01.pid-Des.default.path.home=/tmp/blubb -Des.default.path.logs=/var/log/mylogs/'
}

that would allow the user to specify whatever they like it to be.
OpenBSD users are usually used to those service flags.

Alternatively, as I did, I configured it for the instances in Hiera yaml:
elasticsearch::instances:
  'es01':
    datadir: '/var/elasticsearch/es-data-es01/'
    logging_level: 'DEBUG'
    config:
      path:
        conf: '/etc/elasticsearch/es01/'
        logs: '/var/log/elasticsearch/'

Both ways should work fine.

electrical commented:
Left a few comments.
Could you also test it with ES 2.0 ?

Thanks!

buzzdeee commented:
OpenBSD only currently comes with 1.7.X release of elasticsearch. I'm aware for logstash, a 2.X is in the queue to get added, but I don't know for ES, maybe also soon in the future.
Since I run OpenBSD -current mostly everywhere, whenever ES gets updated to 2.X and I run into trouble with the module, I'll likely open a new PR with patch ;)

cheers,

electrical commented:
Cool. Thank you for the clarification.
I assume that we don't need to setup any special repositories then ?
I'll kick off jenkins for the current things.
I don't have an openbsd vagrant box to run the tests against but will be doing that at some point so I can be sure not to break things later :)

electrical commented:
jenkins, test it

buzzdeee commented:
No extra repo required, it's maintained in the ports tree, and packages available on  all the OpenBSD mirrors.

Note: I use this @home, or just start looking into it, learning. I'm OpenBSD only, and run (more or less current) OpenBSD -current here. Similar aproach to the OpenBSD -current, I run with Puppet modules. I've r10k setup, to fetch all modules I use from its git repo master branch. I don't use them from the forge. When things suddenly start to break for me (I had this in the past with a few modules already) be sure I'll at least complain, or even provide a PR ;)

thanks,
Sebastian

buzzdeee commented:
BTW, elasticsearch package for OpenBSD was just updated to 2.0 today ;). So will likely test with that new version in a few days.

buzzdeee commented:
meanwhile tested with ES 2.0.0, and even ES 2.1.0. Basic installation, configuration and service management works as advertized. 
However, plugin management doesn't work with those 2.X version anymore, but I think that is not unique to OpenBSD ;)

electrical commented:
jenkins, test it

electrical commented:
@buzzdeee plugin install for 2.x is indeed broken, but working on a PR to fix that.

electrical commented:
@buzzdeee sorry i haven't looked at this PR anymore. been busy fixing the plugin installation for 2.x which is finally in.
I'll be looking at this PR very soon.

buzzdeee commented:
@electrical thanks for getting back. I just rebased, and tried again to manage plugins but got odd failures about missing dependencies. I'll look into it in the next few days, when I figure how to fix that, and make the plugins management work, I'll get back here.

electrical commented:
@buzzdeee sounds good. Feel free to ping me at anytime if you need any help.

buzzdeee commented:
so, got plugin installation to work on OpenBSD.
Mainly see my commit message, what I did to it. I kept it as a separate commit to ease working on it based on feedback.

I just saw, and maybe that may be preferred:
  * instead of that ugly additional parameter, I could define the plugin_dir like  ${elasticsearch::params::homedir}/plugins, as it is done in scripts.pp for the
scripts directory, Alternatively I could directly use the ${elasticsearch::params::plugindir} 
and hand that over to the native plugin type from plugin.pp.

Then maybe even remove the defaultto of the plugin_dir parameter of the native plugin type?

That should then also give similar/same results on Linux as well as on OpenBSD.

I think proposed approach would make more sense, so let me know if you prefer that, and it should be easy to change for me.

electrical commented on lib/puppet/provider/elasticsearch_plugin/plugin.rb:
```diff
@@ -90,7 +98,12 @@ def destroy
-      version = es('-v') # ES 1.x
+      os = Facter['osfamily'].value
+      if os == 'OpenBSD'
+        version = pkg_info('elasticsearch').lines.first
```
Don't think this special case is required since getting the version via the elasticsearch binary should work on any platform right?

buzzdeee replied:
running the plain binary I get:
# /usr/local/elasticsearch/bin/elasticsearch --version
You must set the ES_CLASSPATH var

So I would have to run it like this:
ES_INCLUDE="/etc/elasticsearch/elasticsearch.in.sh" /usr/local/elasticsearch/bin/elasticsearch --version

to get the version output,
so I would have to special case it in any way. 
If you prefer, I can look into setting the environment variable, before attempting to run the command, instead of using pkg_info. Shouldn't be too complicated.

electrical replied:
That's weird it doesn't happen on the debian/ubuntu/centos based machines.
elasticsearch binary should be handling the elasticsearch.in.sh automatically.

electrical replied:
Ah. we don't look in /usr/local/elasticsearch/bin by default at https://github.com/elastic/elasticsearch/blob/master/distribution/src/main/resources/bin/elasticsearch#L79-L95

Would be nice to have that fixed in the actual binary as well.
For the time being you can indeed set the ENV variable before running the command.
At least that way the local stays the same.

electrical commented on manifests/plugin.pp:
```diff
@@ -75,6 +81,7 @@
     $source      = undef,
     $proxy_host  = undef,
     $proxy_port  = undef,
+    $plugin_dir  = undef,
```
We can pass the plugin_dir from the params class here so we get the right one depending the distro. This way the user doesn't have to specify it them selves.
Also the plugin dir is set via the main class and passed on internally setting different variables ( including config file )

electrical commented on manifests/package.pp:
```diff
@@ -38,19 +38,22 @@
-      command => "mkdir -p ${elasticsearch::package_dir}",
-      creates => $elasticsearch::package_dir,
-    }
+    $package_dir = $elasticsearch::package_dir
```
the package dir is used for puppet:// based plugins so its fairly mandatory to be created, otherwise it will fail.
See https://github.com/elastic/puppet-elasticsearch/blob/master/manifests/plugin.pp#L115

buzzdeee replied:
see below, its not entirely removed, its only guarded by a if $package_dir.
For OpenBSD, I've set it to undef, see in the params.pp.
Since I used the packages that come with the distribution, I don't have a need for the package_dir variable to be set. However, people should still be able to override it as parameter if really needed.

electrical replied:
In this case the package_dir var is used by the plugin define as well to have a temp location to place the plugin when it comes from a puppet:// resource for offline installations, so i think it needs to be re-implemented for OpenBSD as well.

buzzdeee replied:
ah, that I haven't seen, will look into it, and find a suitable default value.

electrical commented:
Left 1 comment, looks good further!

buzzdeee commented:
I think I addressed all your points, also got the plugin installation working. Had to export two environment variables, and resetting them afterward, to not pollute the rest of the puppet run with it.
Let me know if there is anything else.

electrical commented:
@buzzdeee thank you! I'll try to review this today or tomorrow and let jenkins do its magic in the meantime to ensure it doesn't break existing stuff.

electrical commented:
jenkins, test it

electrical commented:
Code looks good! Waiting for Jenkins to report back and then i can merge it.
I'll have to look for a vagrant box of OpenBSD to so i can run tests on it to ensure i don't break it in the future.

electrical commented on manifests/plugin.pp:
```diff
@@ -77,7 +77,7 @@
 ) {
 
-  include elasticsearch
+  require elasticsearch
```
This seems to have caused a dependency cycle.

```
 Error: Could not apply complete catalog: Found 1 dependency cycle:
  (Augeas[defaults_es-01] => Elasticsearch::Service::Init[es-01] => Elasticsearch::Service[es-01] => Elasticsearch::Instance[es-01] => Class[Elasticsearch] => Elasticsearch::Plugin[mobz/elasticsearch-head/1.0.0] => Elasticsearch_plugin[mobz/elasticsearch-head/1.0.0] => Elasticsearch::Plugin[mobz/elasticsearch-head/1.0.0] => Class[Elasticsearch])
```

Don't think we need the require for accessing the variable ( plugindir ) since it always worked with the restart_on_change one.

Also in https://github.com/elastic/puppet-elasticsearch/blob/master/manifests/init.pp#L411 there is already an explicit dependency for it.

buzzdeee replied:
Thanks Mr. Jenkins for spotting that ;)

I tested it for me, and did not ran into recursion issues, therefore kept it.
but I also don't have to deal with that augeas thingie bingie for that defaults file on OpenBSD.

reverted it back to 'include', since the include also seems to be fine for me.

electrical replied:
This hasn't changed yet by the looks of it and is still causing the test failure.

buzzdeee replied:
crap, I changed it, but then squashed/reordered commits and it reappeard without /me noticing.
Now it's back as it should be.

electrical replied:
hehe no worries. can happen.
Will let Jenkins do its thing again. thanks!

electrical commented:
jenkins, test it

electrical commented:
jenkins, test it

electrical commented:
Hmm, don't know why jenkins didn't trigger. stupid thing. lets try it again

electrical commented:
jenkins, test it

electrical commented:
jenkins, test this

electrical commented:
jenkins, test it

electrical commented:
everything looks good, just want to let CI run it a bit more. found some weird testing things.

elasticsearch-release commented:
Can one of the admins verify this patch?

buzzdeee commented:
for your convenience, rebased

electrical commented:
Thank you for the rebase. Made some changes to make testing more stable.
Will let Jenkins do some more work and merge it tomorrow.

electrical commented:
jenkins, test it

electrical commented:
jenkins, test it

electrical commented:
@buzzdeee I'm afraid a rebase is needed. sorry.

buzzdeee commented:
no worries, rebased yesterday evening, still seems to be fine for me (:

electrical commented:
jenkins, test it

electrical commented:
Sorry, this occidentally dropped of my radar last week. Hoping jenkins will do its thing this time and i'll merge it and release next week.

electrical commented:
jenkins, test it

buzzdeee commented:
no worries (:
